core/dbt/clients/registry.py

```diff
@@ -26,20 +34,27 @@ def _get_with_retries(path, registry_base_url=None):
 
 def _get(path, registry_base_url=None):
     url = _get_url(path, registry_base_url)
+    if url in REGISTRY_CACHE.keys():
+        fire_event(RegistryCacheHit(url=url))
+        return REGISTRY_CACHE[url]
     fire_event(RegistryProgressMakingGETRequest(url=url))
     resp = requests.get(url, timeout=30)
     fire_event(RegistryProgressGETResponse(url=url, resp_code=resp.status_code))
     resp.raise_for_status()
 
+    json_response = (
+        resp.json()  # if this fails, it should raise requests.exceptions.JSONDecodeError and trigger retry
```

Comment:
One observation here is that if we have standards about not just that the response is JSON, but that the JSON is constructed in a very specific way, we would need those logic checks here.  

e.g. the `def package(name, registry_base_url=None):` block makes some assumptions about the structure of the json itself that `resp.json()` won't catch if the response is JSON, but not the *right* JSON.

In the case of separate invocations of deps, we would still, I would imagine, want to retry if the cache was empty -- so I am gathering  `url = _get_url(path, registry_base_url)` is going that -- but I would imagine the logic we might want to employ is to only retry if the cache is empty (e.g. first package in the invocation of `dbt deps`)

Reply:
> not the right JSON. My hypothesis is that ~60% of the current failures we see in Cloud logging, that's exactly what's happening. We get a JSON back, just not the JSON we want, hence the failure with the NoneType in the `package()` call.

@barberscott That's illuminating!

@emmyoop It sounds to me like we might need, in addition to validating whether the JSON is a valid list/dict (#4982), to be able to either:
- Pass a set of "expected" keys into `_get` (e.g. `versions`, `namespace`, `name`, `downloads`, ...), and have it raise a `ContentDecodingError` if a valid-but-incorrect/incomplete JSON body is returned
- Move more validation logic, such as [`RegistryPackageMetadata.from_dict()`](https://github.com/dbt-labs/dbt-core/blob/204d53516a08a619ebea69b68aa8138dd1814f54/core/dbt/deps/registry.py#L56-L58), from `deps/registry.py` to `clients/registry.py`, within the methods where we can retry those requests. It's less-clean code, and worse separation of concerns... but it's also an acknowledgment that these concerns are not easily separated.

The appeal of the change in this PR is that it narrows us down to making only two types of requests: `api/v1/index.json` and `api/v1/{packagename}.json`. Some cluttering of the code for two doesn't feel quite as costly as it would for many.

Reply:
> In the case of separate invocations of deps, we would still, I would imagine, want to retry if the cache was empty

Yeah, this cache won't be persisted _at all_ between separate invocations. I feel even better saying that after having switched out my janky "global" for the Python stdlib approach.